### PR TITLE
New Adapter: Revcontent

### PR DESCRIPTION
**Pull request by jcamp-revc**
Hello!

I'm looking to add support for Revcontent ORTB to prebid-server. We really just need a [generic adapter](https://github.com/prebid/prebid-server/issues/1429), but from the discussion I don't think that's a thing yet.

I'm fairly new to go and this project, so just let me know what I need to fix. :)

### CODE REVIEW

File: adapters/revcontent/revcontent.go
Comment (SyntaxNode) [inline adapters/revcontent/revcontent.go:14]: Nitpick: We prefer the name `adapter`, as the object is referenced by the `adapters.Bidder` interface outside of this package. There's not need to duplicate the name of the package or export it (start with an uppercase letter).

File: adapters/revcontent/revcontent.go
Comment (SyntaxNode) [inline adapters/revcontent/revcontent.go:27]: This method is really just calling the `makeRequest` and adds no new logic. I recommend moving the content of the `makeRequest` method here, such as:

```
func (a *adapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
	reqBody, err := json.Marshal(request)

	if err != nil {
		return nil, []error{err}
	}

	headers := http.Header{}
	headers.Add("Content-Type", "application/json;charset=utf-8")

	req := &adapters.RequestData{
		Method:  "POST",
		Uri:     a.endpoint,
		Body:    reqBody,
		Headers: headers,
	}
	return []*adapters.RequestData{req}, nil
}
```

File: adapters/revcontent/revcontent.go
Comment (SyntaxNode) [inline adapters/revcontent/revcontent.go:65]: Nitpick: In Go, it's preferable for error messages to be start a lower case letter and not end with punctation. 
`"unexpected status code: %d"`

File: adapters/revcontent/revcontent.go
Comment (SyntaxNode) [inline adapters/revcontent/revcontent.go:82]: It's not worth counting the bids to use an exact number here, IMHO. The effort of counting likely won't offset the small memory performance increase by specifying a guess. Set this to the number of bids your server sends back on average.

File: usersync/usersyncers/syncer_test.go
Comment (SyntaxNode) [inline usersync/usersyncers/syncer_test.go:112]: Please add your adapter alphabetically to this list. Although to be fair, it seems to be in need of sorting.

File: adapters/revcontent/revcontent.go
Comment (SyntaxNode) [inline adapters/revcontent/revcontent.go:112]: Your bidder doesn't support video, so this case doesn't seem possible.

File: adapters/revcontent/revcontent.go
Comment (SyntaxNode) [inline adapters/revcontent/revcontent.go:123]: It would be ideal to use a response from your bidder to set the media type. Does your bidder include any ext values to use in that determination? If not, what would your bidder respond back with if the impression is for both banner and native?

**Review by SyntaxNode — COMMENTED** [review]
Looks great. Thank you for your contribution. I have one comment on determining bid type and the rest are all syntax nitpicks. Test coverage looks great.

Seems like you followed the new docs. I'm curious how that worked out for you or if you recommend any improvements.

**Comment by jcamp-revc** [comment]
Thanks for the feedback. I think this should all be addressed in the latest commit. The new docs were very helpful, thank you!

File: adapters/revcontent/revcontent.go
Comment (mansinahar) [inline adapters/revcontent/revcontent.go:84]: Looks like this function is unused. Can you please remove it?

File: adapters/revcontent/revcontenttest/params/race/banner.json
Comment (mansinahar) [inline adapters/revcontent/revcontenttest/params/race/banner.json:2]: If your adapter doesn't support any parameters then you don't need these params files. You can remove these.

**Review by mansinahar — COMMENTED** [review]
Except for the following minor comments, LGTM!

**Comment by jcamp-revc** [comment]
> Except for the following minor comments, LGTM!

Thanks, these should be addressed now.

File: static/bidder-params/revcontent.json
Comment (SyntaxNode) [inline static/bidder-params/revcontent.json:1]: Nitpick: Please remove the leading line break.

File: adapters/revcontent/revcontent.go
Comment (SyntaxNode) [inline adapters/revcontent/revcontent.go:14]: This doesn't need to be exported and should therefore start with a lower case letter. Please rename to `adapter`.

File: adapters/revcontent/revcontent.go
Comment (SyntaxNode) [inline adapters/revcontent/revcontent.go:90]: Nice. Thank you.

File: static/bidder-params/revcontent.json
Comment (SyntaxNode) [inline static/bidder-params/revcontent.json:7]: Does your adapter truly have no bidder parameters? If so, not a problem.

If you do have bidder parameters but aren't defining them here because you pass the OpenRTB request exactly as is, please consider adding them here to allow PBS to enforce the parameters are provided and prevent outgoing traffic to your bidding server if the request is invalid.

**Review by SyntaxNode — COMMENTED** [review]
Looks great with ideal test coverage. Just a few more nitpicks and a question.

File: static/bidder-params/revcontent.json
Comment (jcamp-revc) [reply static/bidder-params/revcontent.json:1]: whoops.

File: static/bidder-params/revcontent.json
Comment (jcamp-revc) [reply static/bidder-params/revcontent.json:7]: Do you mean things like `bidfloor`? Then yes, I suppose I'll have to look at this. Thanks!

Comment (jcamp-revc) [reply static/bidder-params/revcontent.json:7]: We really just need to look at:
(request.App.name OR request.Site.domain) AND request.Device, and everything else is optional on our end.

Do we need to enforce these in a check following this example: https://github.com/prebid/prebid-server/blob/master/adapters/pubnative/pubnative.go#L80-L88

Or is there a way to use the params.json to check those? We don't currently need any ext params or anything.

Comment (SyntaxNode) [reply static/bidder-params/revcontent.json:7]: I wouldn't say you need to, but I do highly recommend it. The earlier we can disqualify a bid request the less resources we all have to use to accomodate it. In this case, PBS hosts wouldn't need to reach out to your bidding servers for requests which are known to be invalid.

This must be coded in the adapter, as you linked to for pubnative. The params.json file is just to check bidder extensions not define within the OpenRTB spec.

Thank you for the details. This json file is fine as is. I do encourage you to add the checks you mentioned along with new json tests cases to cover the new logic.

**Comment by SyntaxNode** [comment]
Removing stale label to end of year holidays.

**Comment by jcamp-revc** [comment]
Thanks, I'll get those checks added and then hopefully we can merge soon.

File: adapters/revcontent/revcontent.go
Comment (SyntaxNode) [inline adapters/revcontent/revcontent.go:56]: This check is performed by the "core" and isn't necessary here. See: https://github.com/prebid/prebid-server/blob/a709baa4f842b11d4711b642c68ee245639dde07/endpoints/openrtb2/auction.go#L336

**Review by SyntaxNode — COMMENTED** [review]
Thank you for adding the validation. One the checks is already covered by "core" code and doesn't need to be repeated.

File: adapters/revcontent/revcontent.go
Comment (jcamp-revc) [reply adapters/revcontent/revcontent.go:14]: sorry, not sure how that got missed. Fixing.

**Comment by SyntaxNode** [comment]
Thank you for the changes. Looks like it's failing on `app_and_site.json` and that's ok. The adapter json unit tests bypass the core checks. You can resolve by removing that test. We have coverage elsewhere.

**Comment by jcamp-revc** [comment]
> Thank you for the changes. Looks like it's failing on `app_and_site.json` and in retrospec that's expected. The adapter json unit tests bypass the core checks. You can resolve by removing that test. We have coverage elsewhere.

That's what I get for not running the tests, even on a simple change. Sorry about that.

**Review by SyntaxNode — DISMISSED** [review]

**Comment by SyntaxNode** [comment]
LGTM. Thank you. I'll seek a second review tomorrow and there's a very good chance this can be incorporated into the next release. Before we can merge, we just need you to open the documentation PR. For details, please see the bottom of the new adapter docs: https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#user-documentation

**Review by bsardo — DISMISSED** [review]
LGTM

**Comment by SyntaxNode** [comment]
@jcamp-revc We're ready to merge in once you have a PR to update your adapter docs.

**Comment by jcamp-revc** [comment]
Thanks, [PR is opened](https://github.com/prebid/prebid.github.io/pull/2614).

**Comment by SyntaxNode** [comment]
@jcamp-revc Thank you.

I noticed that "Revcontent will provide you with a custom endpoint that will contain your account details". With this in mind, does the default URL provided in this PR work? .. or should this adapter be disabled by default until hosts work with Revcontent to set it up?

**Comment by jcamp-revc** [comment]
I tried to disable it, but was having trouble with getting tests to pass. Let me try again.

File: config/config.go
Comment (bsardo) [inline config/config.go:982]: The second parameter here should be a boolean instead a string.

Comment (jcamp-revc) [reply config/config.go]: Thanks. Sorry about that.

Comment (SyntaxNode) [reply config/config.go]: No worries. Thanks for fixing it. There's just a pesky merge conflict now.

Comment (jcamp-revc) [reply config/config.go]: Is that for me to fix, or you?

Comment (SyntaxNode) [reply config/config.go]: That's for you to fix. Please merge in the master branch and resolve the conflict. Looks like a super minor issue with the user sync tests.

**Review by SyntaxNode — APPROVED** [review]

**Review by bsardo — APPROVED** [review]

**Comment by terrapop** [comment]
Does this adapter work? Seems it sends out ad requests, but we don't get any response from Revconent at all. Our setup looks ok, and we used the correct values for the API key and account ID.

**Comment by SyntaxNode** [comment]
@terrapop If you don't receive a response from @jcamp-revc here, please reach out to their support email address developers@revcontent.com.

We approved the adapter based on the sample requests included in this PR. We don't have test accounts or test placements with Revcontent to be able to verify further. It's the responsibility of bidders to perform end to end testing on their submissions.

**Comment by jcamp-revc** [comment]
> Does this adapter work? Seems it sends out ad requests, but we don't get any response from Revconent at all. Our setup looks ok, and we used the correct values for the API key and account ID.

@terrapop  Hi. I'm happy to take a look at this. Can we move over to email at jon.camp@revcontent.com and I'll help you look into it.

**Pull request merged**
